feat(Menu): add app-launcher mockup

kmcfaul opened the pull request:
<!-- What changes are being made? Please link the issue being addressed. -->
**What**: Closes #6018

In this PR:
- Adds mockup of ApplicationLauncher (using Menu components) to the Composable Menu demo page
- Changes typing on MenuItem `component` property to allow router `Link`
- Sets `id` on `MenuItem` if given `itemId` (manually setting `id` directly should override `itemId` if both are passed)

The demo covers most of the examples shown in ApplicationLauncher. The CSS overall is similar but there are slight differences in spacing. I think the next step would be to decide how we want to handle those changes, and decide if we want to leave this as a demo or make a new AppLauncher wrapper component for Menu.

patternfly-build commented:
PF4 preview: https://patternfly-react-pr-6175.surge.sh

CooperRedhat commented on packages/react-core/src/demos/ComposableMenu.md:
```diff
@@ -869,14 +868,240 @@ MenuWithFlyout = () => {
-      isVisible={isOpen}
-      popperMatchesTriggerWidth={false}
-    />
+    <Popper trigger={toggle} popper={menu} isVisible={isOpen} popperMatchesTriggerWidth={false} enableFlip={false} />
```
Why no `enableFlip`?

kmcfaul replied:
Ah that was just me forcing the menu to expand downwards (since its last on the page). I can re-enable it.

CooperRedhat reviewed (approved):
LGTM

kmcfaul commented:
Removed `itemId` setting `id` on MenuItem - `itemId` is specific to return calls and not necessarily tied to `id`. I could re-add this if others think it's worth having `itemId` set `id` automatically (without needing to pass both `itemId` and `id`).

mcarrano reviewed (commented):
This looks good @kmcfaul . The only problem I see is that the filtering does not seem to be working. For whatever I type I get "No results found."

nicolethoen reviewed (commented):
I'm noticing a weird keyboard interaction:

1.  Use `enter` to open the menu
2. Focus moves to the search
3. `tab` exits the search and moves to the menu items where I can navigate no problem with arrow keys
4. `tab` does not move the focus to the next navigable item on the page that i'd expect. It moves it off the page to the browser address bar. 

It seems that the keyboard navigation is being confused by the menu being appended to the end of the DOM rather than between the app launcher toggle and the example's typescript toggle button...

mcoker commented:
This looks pretty good. Some things I noticed:

* Group title bottom padding differs - 8px on app launcher, 0px on menu
  * You can see that by hovering over a menu item and noting how close the background color is to the title
  * app launcher
    * <img width="190" alt="Screen Shot 2021-08-17 at 5 43 58 PM" src="https://user-images.githubusercontent.com/35148959/129810340-2c94825f-9abc-4a0b-bb6d-fe89b35b8e4f.png">
  * menu
    * <img width="378" alt="Screen Shot 2021-08-17 at 5 44 05 PM" src="https://user-images.githubusercontent.com/35148959/129810348-92f413db-9cdf-406d-af01-d47493119d32.png">


* Search box top/bottom padding differs - 16px top/bottom on app launcher, 24px top/bottom on menu
* We impose height/width/alignment restrictions on app launcher icon. No restrictions on the menu icon.
  * The app launcher icon box is 24x24px, then we size down the largest edge (of the icon/image/whatever the user passes to the icon box) to 24px and allow the other edge to scale down.
  * It's up to the user to use the right/consistent sized image in a menu item icon. This is probably fine if they're using react-icons, not so much if using custom icons that may be sized differently.

@mcarrano do we want to update any of this in the menu component CSS to be more inline with app launcher? The app launcher should represent what we'll see in other dropdown type components.

kmcfaul commented:
@mcarrano Fixed the search, it was failing to filtered properly because options lacked `id` (after I removed `itemId` setting `id`). 

@nicolethoen Appended the popper to a container div to get the proper tab order. I could also remove the popper but then the example container grows/shrinks when the menu is opened/closed. Keyboard interaction is a bit odd for this example specifically, because menu search is outside of the `ul` options and isn't a part of the built in navigation (same with menu footer, you can see in the Menu examples). I removed `Tab` closing the menu for this example to allow access to the options. I can try playing with the interaction a bit more, I can maybe build arrow navigation for the input in the demo code.

mcoker commented:
@kmcfaul @mcarrano re: [this comment](https://github.com/patternfly/patternfly-react/pull/6175#issuecomment-900682527), I looked closer and I think it's just a markup issue:

* `<MenuList>` shouldn't wrap the whole menu - it should only wrap `<MenuItems>`
* Unless the `<Divider>` is a direct child of a `<MenuList>` (it's used to separate items in a single list), it should be an `<hr>` - doesn't need the `component` prop. Looks like all of the dividers in this menu should be `<hr>`s since they separate groups of things, not items.

Here is an example of the search box markup just showing the elements for the search box + divider - https://pf4.patternfly.org/components/menu#with-filtering

<img width="333" alt="Screen Shot 2021-08-18 at 1 34 17 PM" src="https://user-images.githubusercontent.com/35148959/129953256-526aef28-17ff-4e59-abc1-0df3339c77a3.png">

And here is an example of the favorites markup for what to put in `<MenuContent>` - https://pf4.patternfly.org/components/menu#with-favorites

<img width="403" alt="Screen Shot 2021-08-18 at 1 31 42 PM" src="https://user-images.githubusercontent.com/35148959/129953365-19036108-672d-4173-8560-78a6e42e95e8.png">

kmcfaul commented:
@mcoker Updated the markup & dividers

mcoker commented on packages/react-core/src/demos/ComposableMenu.md:
```diff
@@ -869,14 +868,259 @@ MenuWithFlyout = () => {
+      </MenuInput>
+      <Divider />
+      <MenuContent>
+        <MenuList>
```
This needs to wrap the `<MenuItem>`s. I'm assuming it would wrap `{filteredFavorites}` a few lines below this?

FWIW `<MenuList>` generates a `<ul>` element. The only valid children of a `<ul>` are `<li>` elements (what the `<MenuItem>` generates here). And the inverse is similar, `<li>` elements can only be a child of `<ul>` or `<ol>`.

kmcfaul replied:
`filteredFavorites` and `filteredItems` are both collections of `MenuItem` / `MenuGroup` so it should be good.

mcoker commented on packages/react-core/src/demos/ComposableMenu.md:
```diff
@@ -869,14 +868,259 @@ MenuWithFlyout = () => {
+          {filteredFavorites.length > 0 && (
+            <MenuGroup key="favorites-group" label="Favorites">
+              {filteredFavorites}
+              <Divider key="favorites-divider" />
```
This should go between `<MenuGroup>`s.

kmcfaul replied:
If it's outside of `MenuGroup` that would place it directly under `MenuList`, would it need the `component='li'` in that case then?

nicolethoen commented:
I like the updates to the keyboard interaction. It's not ideal but I think the issue of a11y when it comes to items appended to the DOM outside of their apparent parent is something we might have to solve on a larger scale as well.

Thumbs up from me once you address Coker's comments! :)

mcoker reviewed (dismissed):
LGTM! Cool to see this come together!!

redallen commented on packages/react-core/src/components/Menu/MenuItem.tsx:
```diff
@@ -102,7 +102,10 @@ export const MenuItem: React.FunctionComponent<MenuItemProps> = ({
     disableHover
   } = React.useContext(MenuContext);
-  const Component = component || to ? 'a' : 'button';
+  let Component = !component && to ? 'a' : 'button';
+  if (component) {
+    Component = component as any;
+  }
```
You can write `const Component = (!component && to ? 'a' : 'button') as any;`

kmcfaul replied:
Will that preserve a defined `component` though? The first statement will fall through to setting `button` as the component. But if a `component` is given it should override that assignment.

redallen replied:
Oh sorry I missed that logic, you could do `const Component = (component || to ? 'a' : 'button') as any;` in that case.

tlabaj replied:
I also found this a little confusing.  Also why don't we just default component to `button` on line 80?

kmcfaul replied:
I was trying to preserve the logic that let `to` set `component` to `a`. 

`component || to ? 'a' : 'button'` wouldn't keep a given `component` either since it's a if-else.

I could use `(to ? 'a' : component) as any` and default `component` to `button`.

redallen commented on packages/react-core/src/demos/ComposableMenu.md:
```diff
@@ -4,6 +4,8 @@ section: demos
 beta: true
 ---
 
+import { Link } from '@reach/router';
```
Conflicts in this file.

kmcfaul replied:
What do you mean?

kmcfaul commented:
@redallen @tlabaj 

Updated the menuitem logic for `component`. It looks like something got updated for the menu/popper CSS though, the width is now the full page? @mcoker @mattnolting

kmcfaul commented:
Styling will get another quick update for the menu width, but shouldn't have any changes on the react-side for it

redallen commented on packages/react-core/src/components/Menu/MenuItem.tsx:
```diff
@@ -102,7 +102,7 @@ export const MenuItem: React.FunctionComponent<MenuItemProps> = ({
     disableHover
   } = React.useContext(MenuContext);
-  const Component = component || to ? 'a' : 'button';
+  const Component = (to ? 'a' : component) as any;
```
I like the readability!

mcoker reviewed (dismissed):
LGTM!

CooperRedhat commented:
https://user-images.githubusercontent.com/87719785/130282146-747e46a3-0fa9-4eb5-bb23-97b156635eaa.mov

Just saw this on surge

kmcfaul commented:
@CooperRedhat Ah good catch, I need to update the filtering logic since the HTML structure has changed a bit since it was initially built. Will have an update out soon.

patternfly-build commented:
Your changes have been released in: 
 - @patternfly/react-catalog-view-extension@4.12.52
 - @patternfly/react-code-editor@4.3.38
 - @patternfly/react-console@4.11.20
 - @patternfly/react-core@4.152.0
 - @patternfly/react-docs@5.21.10
 - @patternfly/react-inline-edit-extension@4.7.61
 - demo-app-ts@4.118.3
 - @patternfly/react-log-viewer@4.4.17
 - @patternfly/react-table@4.29.54
 - @patternfly/react-topology@4.9.58
 - @patternfly/react-virtualized-extension@4.9.27

Thanks for your contribution! :tada:
